### ISSUE TITLE
Rewrite "RSpec"y Context/Should tests into before_all/test

### ISSUE THREAD

**Issue by mbarnett** [ualbertalib/jupiter#417]
It's a bit disappointing that this split in style crept in at all, but I've been seeing a lot of flakiness with flapping tests, surprising and unintuitive cases of repeated work slowing down tests, etc, stemming from all of the crufty magic this DSL performs (quick: describe the concrete behaviour of nested `context` blocks with multiple `setup`s)

(example weirdness I've seen from this nested context stuff: I've had assertions fail on the line _after_ a `binding.pry` I inserted to find out why it was failing, without the `binding.pry` activating first. This stuff is really, really fragile and hard to work with)

In contrast the `before_all`/`test 'test name' do` behaviour is straightforward and easy to reason about. I think we should standardize on it.

**Comment by pgwillia** [comment]
- [ ] remove contexts
- [ ] change all before_all to setup
- [ ] change all should to test
- [ ] create fixtures for reused setup

**Comment by murny** [comment]
Some extra thoughts. 

- If you can remove all context/should blocks, then we can replace `gem shoulda` for `gem shoulda-matchers` (as gem shoulda contains both shoulda-matchers and shoulda-context (context being the one that gives should/context blocks, and shoulda-context is whats holding back shoulda and isn't maintained so big win)


- `after_all/before_all/before(:all)` is actually from a gem called https://github.com/jeremyevans/minitest-hooks (`setup` is a bit different as this runs before every test). Is this needed? Not really, since minitest is just ruby :tm: you can just remove those blocks and leave the code like:

```ruby
require "test/helper"

class TestBlah < TestWhatever
  @@expensive_fedora_item_creation ||= do_expensive_item_creation

  def test_something
    assert_equal 'Cool title', @@expensive_fedora_item_creation .title
  end
end
```
And places where we using it to wipe out the entire fedora database in test_helper.rb:

```ruby
 # clean ActiveFedora at the end of all tests in a given test class
  def after_all
    # we clean fedora here
  end
```

Minitest  provides native helpers for this as well called:  `before_suites/after_suites` 

So do we need this gem? Probably not. More info here about why minitest doesn't have a before_all/etc:  https://github.com/seattlerb/minitest/issues/61

**Comment by pgwillia** [comment]
Here are some things I've noticed:

* `before_all` is a hook provided by minitest-hooks and a common `after_all` hook in the test_helper.rb to clean up -- trying to move these blocks to use class level variables isn't working like I'd like
* `def setup` vs `setup do` - `setup do` [is a callback from rails](https://github.com/rails/rails/blob/d66e7835bea9505f7003e5038aa19b6ea95ceea1/activesupport/lib/active_support/testing/setup_and_teardown.rb#L28), which runs before [minitest](https://github.com/seattlerb/minitest/blob/master/lib/minitest/test.rb#L158) `TestCase#setup` Do we have a preferred style?
* biggest candidates for reuse right now is community/collection -- can't fixture them as is because their not ActiveRecord objects (error because the default fixture cleanup can't find those tables) Is there something obvious to use instead of fixtures for these?  FactoryGirl?

**Comment by pgwillia** [comment]
Candidates for refactor
[duplicates.txt](https://github.com/ualbertalib/jupiter/files/1933813/duplicates.txt)

**Comment by pgwillia** [comment]
Feeling a little stuck so let's talk about this.

1. The setup for tests.
 * currently use minitest hooks (`before_all`/`after_all` which we have some experience of hiding exceptions)
 * use constant block (is there a way that we can use `file_fixture` in the block? Won't work for me :-1:  )
```
SETUP = begin
 @@variable = expensive_setup
end
```
 * use memoization/lazy accessor in a setup block maybe with a test_helper method.  By this I mean check to see if the object exists, if it does then use it, otherwise create it.  `Community.where(title: 'Fancy Community').first || Community.new_locked_ldp_object(title: 'Fancy Community', owner: 1).unlock_and_fetch_ldp_object(&:save!)` The added benefit of this is the minitest discussion of randomizing across tests would work. This also feels like we're entering fixtures territory
 * https://gist.github.com/courtenay/7343845 -- define fixtures in a block a la [running-man](https://github.com/technoweenie/running_man) 
 * some test classes contain only one test so expensive setup could happen in `setup do` or the test itself and the problem is punted into the future
2. The teardown for tests
 * currently use minitest hooks (after_all) to clean up Fedora, Redis and Sidekiq
 * move to the first thing done in the setup
 * use `Minitest.after_run` to do cleanup (runs after all tests not the end of all tests in a given test class)
 * overload Minitest.run
```
class MyHeavyTest < Minitest::Test
  def self.run(*args)
    # do stuff before suite
    super
    # do stuff after suite
  end
end
```
 * googling hasn't given me any other suggestions other than massive tests which contain all setup, all tests, and all teardown -- suggestions?
3. Fixtures
 * put in own directory because not ActiveRecord FixtureSet (and table for teardown won't be found)
 * can create a helper method
```
# uses :title as a lookup so must be unique
def locked_ldp_fixture(class_name, options)
  class_name.where(title: options[:title]).first || class_name.new_locked_ldp_object(options).unlock_and_fetch_ldp_object(&:save!)
end
```
 * how to handle `unlock_and_fetch_ldp_object` blocks that contain add_to_path, filesets and the like?
 * https://gist.github.com/avastor/a1f088ca58860eff47a3 -- fixtures (without ActiveRecord)

**Comment by mbarnett** [comment]
If you're feeling super stuck I'd probably just move on? We can incrementally improve the test situation over time.

The set up and fixtures issues probably aren't worth sinking a _ton_ of time into, because if we're looking at a transition off of Fedora soon anyways, we'll get fixtures for free without having to do any kind of bespoke work to square the circle between the fixture infrastructure and the current LockedLDPObject setup. Things like memoization are going to run into huge issues because our tests assume they can mutate the objects without needing to reset them to a known state at the end of the test run -- heading down that road will just lead to an enormous number of tests being rewritten.

Hooking Minitest#run looks like it might be complicated because of the way it's set up to optionally run tests in parallel -- might be worth future investigation, but if we have to live with keeping Minitest hooks for now, I think we've still made progess?

**Comment by pgwillia** [comment]
From flay test > duplicates.txt

- [x] JOHNDOE_AUTH_HASH

2) won't fix

3) won't fix

- [x] removed second @item1

5) won't fix

6) won't fix

7) won't fix -- I think I like that these tests test one thing -- faster debugging

8) won't fix

- [ ] Random Item Fixture
  test/models/file_set_test.rb:35
  test/models/item_test.rb:331

10) won't fix

11) won't fix

- [ ] Fancy Item Fixture
  test/controllers/file_sets_controller_test.rb:13
  test/integration/sitemap_test.rb:12
  test/integration/sitemap_test.rb:43

- [ ] Nice Community and Collection fixture
  test/controllers/admin/collections_controller_test.rb:5
  test/controllers/collections_controller_test.rb:5

14) won't fix

- [ ] Books Community and Collection fixture
  test/controllers/items/files_controller_test.rb:6
  test/system/deposit_item_test.rb:5

16) won't fix

- [ ] Books Community and Collection fixture
  test/controllers/items/draft_controller_test.rb:5
  test/models/draft_item_test.rb:5

18) not sure similar attributes except for licence/visibility
  test/models/draft_item_test.rb:124
  test/models/draft_item_test.rb:150

19) won't fix

20) won't fix

21) won't fix

22) won't fix

23) won't fix

- [ ] Fancy Community & Collection fixture, maybe Fancy Item fixture
  test/system/item_show_test.rb:15
  test/system/item_show_test.rb:36

25) won't fix - maybe Item for deletion fixture?
  test/controllers/admin/collections_controller_test.rb:93
  test/controllers/admin/items_controller_test.rb:14

- [ ] Fancy Item with files fixture
  test/controllers/file_sets_controller_test.rb:23
  test/controllers/redirect_controller_test.rb:29

27) won't fix

- [ ] Submitted Draft Article fixture
  test/models/item_test.rb:256
  test/models/item_test.rb:285

29) won't fix

30) won't fix

- [ ] Community and foo Collection fixture, maybe Random Item fixture?
  test/models/file_set_test.rb:32
  test/models/item_test.rb:328

32) won't fix

33) won't fix

34) won't fix

35) won't fix

36) won't fix

37) won't fix

38) won't fix

39) won't fix

40) won't fix

41) JupiterCore::LockedLdpObject random public and private object fixture
  test/models/jupiter_core/search_test.rb:37
  test/models/jupiter_core/search_test.rb:40

42) won't fix
